### ISSUE TITLE
Debugging fnn model

### ISSUE THREAD

**Issue by gabrielrueda** [fani-lab/OpeNTF#162]
Hi @hosseinfani, 
I will update you here on the issue for debugging the result of all 0's in the fnn model.

**Comment by hosseinfani** [comment]
@VaghehDashti 
Gabriel is investigating the fnn's results. Please help Gabriel in this task.

**Comment by gabrielrueda** [comment]
Hi @hosseinfani @VaghehDashti, 
I looked into the code and just want to confirm if using `python main.py -data ../data/raw/dblp/toy.dblp.v12.json -domain dblp -model fnn` are the correct arguments to test the fnn model. 

Also, when I run that, I few graph results are shown:
1. For this one, what do the x and y axis numbers represent:
![image](https://user-images.githubusercontent.com/35504507/176545154-34086096-c6dd-4e2e-bf05-9c09b8aecc08.png)

2. For this one, what is meant by "true positive rate" and "false positive rate"
![image](https://user-images.githubusercontent.com/35504507/176544774-5b9af667-ca60-43bc-ae42-978686f06699.png)

**Comment by gabrielrueda** [comment]
Hi @hosseinfani @VaghehDashti, 
Here is an update. I haven't had much time to investigate this issue lately, and I have also been confused on where to start, so I haven't really made an progress so far. I do have a question though. I was looking the results in the csv file (shown below), and I was wondering what P_2, P_5 and P_10 are, as well as what ndcg_cut is and what map_cut is. 

f0.test.pred.eval.mean.csv:
,mean
P_2,0.1
P_5,0.12000000000000002
P_10,0.12
recall_2,0.06666666666666667
recall_5,0.18333333333333332
recall_10,0.4833333333333333
ndcg_cut_2,0.12262943855309168
ndcg_cut_5,0.16448113815022186
ndcg_cut_10,0.27699669758550327
map_cut_2,0.06666666666666667
map_cut_5,0.11000000000000001
map_cut_10,0.15583333333333335
aucroc,0.4164201183431953


I also opened the contents of f0.test.pred.eval.roc.pkl and it shows two arrays. I believe the first array is the training result and the second array is the validation result. 

RESULT: (array([0.        , 0.01923077, 0.01923077, 0.26923077, 0.26923077,
       0.53846154, 0.53846154, 0.55769231, 0.55769231, 0.57692308,
       0.57692308, 0.61538462, 0.61538462, 0.67307692, 0.67307692,
       0.88461538, 0.88461538, 0.98076923, 1.        ]), array([0.        , 0.        , 0.07692308, 0.07692308, 0.15384615,
       0.15384615, 0.23076923, 0.23076923, 0.30769231, 0.30769231,
       0.38461538, 0.38461538, 0.76923077, 0.76923077, 0.84615385,
       0.84615385, 0.92307692, 0.92307692, 1.        ]))

Training: [0.         0.01923077 0.01923077 0.26923077 0.26923077 0.53846154
 0.53846154 0.55769231 0.55769231 0.57692308 0.57692308 0.61538462
 0.61538462 0.67307692 0.67307692 0.88461538 0.88461538 0.98076923
 1.        ]
 
Validation: [0.         0.         0.07692308 0.07692308 0.15384615 0.15384615
 0.23076923 0.23076923 0.30769231 0.30769231 0.38461538 0.38461538
 0.76923077 0.76923077 0.84615385 0.84615385 0.92307692 0.92307692
 1.        ]

**Comment by hosseinfani** [comment]
Hi @gabrielrueda 

these are metrics to measure the accuracy of the model prediction. Also the *.roc.pkl file (yes, they are train-valid results plus ROC). 
Generally, the more close to one the better meaning the more accurate answer by the model.

There is a nice teaching slides here:
https://ils.unc.edu/courses/2013_spring/inls509_001/lectures/10-EvaluationMetrics.pdf

you can read more here:
https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Offline_metrics

**Comment by hosseinfani** [comment]
@gabrielrueda 
I would suggest looking into the the content of f*.test.pred files. They are object files saved by `torch.save()` and contains the matrix to show the actual prediction of the model for each instance in the input. The size should be |test| * |members|

you can load it by `torch.load()`

**Comment by gabrielrueda** [comment]
Hello @hosseinfani @VaghehDashti, 
Here is an update: I went over the slides about the Evaluation Metrics, and I now understand the f0.test.pred.eval.mean.csv contents. I opened the .pred file and looked at the contents. From my understanding, these are the weights are modified as we train the model, right? 

Also, I took a look at a simple example of an fnn model and tried comparing it to the one in this project. 

First, I think I should look at the init function:
```Python
 def init(self, input_size, output_size, param):
        self.fc1 = nn.Linear(input_size, param['l'][0])
        hl = []
        for i in range(1, len(param['l'])):
            hl.append(nn.Linear(param['l'][i - 1], param['l'][i]))
        self.hidden_layer = nn.ModuleList(hl)
        self.fc2 = nn.Linear(param['l'][-1], output_size)
        self.dp = nn.Dropout(0.5)
        self.initialize_weights()
        return self
 ```
- From my understanding we have a matrix of size input_size x param['l'][0] and another matrix of size param['l'][-1] x output_size, with some other matrices in between the two (hidden layers). Is this right?
- Finally, is param just the teams by member matrix like we use in our collabs program, or is there more data stored in the param variable?

**Comment by VaghehDashti** [comment]
> Hello @hosseinfani @VaghehDashti, Here is an update: I went over the slides about the Evaluation Metrics, and I now understand the f0.test.pred.eval.mean.csv contents. I opened the .pred file and looked at the contents. From my understanding, these are the weights are modified as we train the model, right?
> 
> Also, I took a look at a simple example of an fnn model and tried comparing it to the one in this project.
> 
> First, I think I should look at the init function:
> 
> ```python
>  def init(self, input_size, output_size, param):
>         self.fc1 = nn.Linear(input_size, param['l'][0])
>         hl = []
>         for i in range(1, len(param['l'])):
>             hl.append(nn.Linear(param['l'][i - 1], param['l'][i]))
>         self.hidden_layer = nn.ModuleList(hl)
>         self.fc2 = nn.Linear(param['l'][-1], output_size)
>         self.dp = nn.Dropout(0.5)
>         self.initialize_weights()
>         return self
> ```
> 
>     * From my understanding we have a matrix of size input_size x param['l'][0] and another matrix of size param['l'][-1] x output_size, with some other matrices in between the two (hidden layers). Is this right?
> 
>     * Finally, is param just the teams by member matrix like we use in our collabs program, or is there more data stored in the param variable?

Hello Gabriel,
The .pref file contains the predictions of the model.
Yes, that is correct. 
param is a dictionary containing all of the hyperparameters for the model training. param['l'] is a list containing the number of nodes in each hidden layer.

**Comment by hosseinfani** [comment]
@gabrielrueda 
to simply put it, we have input of skills and through matrix multiplication, we land it to output members.

the number of matrixes are defined in the "param" and the actual matrix is built in "init". 

for instance if we have 5 skills in total and 6 members in total, 

if we have a team with skill s2 s4 and members m1, m3, m5,

then the model should map

[0,1,0,1,0] skill vector

to 

[1,0,1,0,1,0] member vector

through multiple intermidiate matrix multiplications.

but the model is not perfect and may land the skill vector to 

[0.9, 0.1, 0.2, 0.1, 0.8, 0.1]

here you can see that the model could almost hit the m1, m5 but misses m3.

each row of the .pred file is for each team whose skill vector is mapped to the member vector, of course with some errors.

when all rows are zero, it means that the model was not able to hit at least one member of at least one team, which is wierd!

hope my reply helps.

**Comment by hosseinfani** [comment]
@gabrielrueda 
Hi Gabriel,
do you have any update?

**Comment by gabrielrueda** [comment]
Hi @hosseinfani,
I am still working through trying to understand fnn.py. I am currently going through some videos on pytorch, to learn what it's various functions do.

**Comment by hamedwaezi01** [comment]
Hi everyone
I have recently started working on this issue, and I am trying to reproduce the problem. I faced some challenges about my laptop hardware specs, as its memory did not suffice. I increased my swap and it seems to be okay for now. In case I had further updates, I will let you know.

**Comment by hamedwaezi01** [comment]
I stopped reproducing the problem on my system. I just realized reproducing on it is not an option. I want to give some hypotheses. Since we are using sigmoid on the output layer, is it possible that the number of epochs might not be enough (It is 20 here)? I am suggesting it as the sigmoid derivative is very close to zero, and it would affect the learning speed.

**Comment by hosseinfani** [comment]
@hamedwaezi01 
can you explain more?

**Comment by hamedwaezi01** [comment]
I knew that using a sigmoid in hidden layers would lead to very slow learning, but I wasn't sure if it had any effects when applied to the output layer. I just searched more about this case and learned it is not valid.
**Sorry for my late response**

**Comment by hamedwaezi01** [comment]
I ran code several times for dblp (the main one) using different hyper-parameters for model. You can see below the results of evaluation step for the corresponding configs.
There is still another session running where the model has a single hidden layer and I will post them soon.
Also, Let's discuss about the results in the next comments.


------------------------------------------------------------------------
Model Layers: [128, 512]
epochs: 60
fold: 0
Positive Weight: 3.5

,mean
P_2,0.00023478902529013216
P_5,0.00030857986180988806
P_10,0.00032870463540618505
recall_2,0.0001397553721965072
recall_5,0.0004919389101317053
recall_10,0.0010476062699850185
ndcg_cut_2,0.0002271988198319274
ndcg_cut_5,0.00038770062421872477
ndcg_cut_10,0.0006411759326708938
map_cut_2,0.00010341897542541534
map_cut_5,0.0002189500831078614
map_cut_10,0.0003324580654023197
aucroc,0.5
![f0 train_valid_loss](https://user-images.githubusercontent.com/124281176/222865800-3edf5047-96da-4d1c-9306-c2c5f5543d67.png)



------------------------------------------------------------------------
Model Layers: [128, 512]
epochs: 60
fold: 2
Positive Weight: 50

,mean
P_2,0.002783927014154424
P_5,0.0023613067686321845
P_10,0.001978936070302551
recall_2,0.0016502314348963562
recall_5,0.003559848840589437
recall_10,0.005926745824109472
ndcg_cut_2,0.002791517219612629
ndcg_cut_5,0.0032185113877667387
ndcg_cut_10,0.004325443815772982
map_cut_2,0.0013086693052480932
map_cut_5,0.0019327236272295631
map_cut_10,0.002305026614746883
aucroc,0.6084372102614889
![f2 train_valid_loss](https://user-images.githubusercontent.com/124281176/222865997-c63c8e04-f689-4061-a6fb-f3b6082cac35.png)


------------------------------------------------------------------------
Model Layers: [128, 512]
epochs: 60
fold: 1
Positive Weight: 300

,mean
P_2,0.0030187160394445564
P_5,0.0024820554102099684
P_10,0.001985644328167981
recall_2,0.0017575635607432738
recall_5,0.00369177791194294
recall_10,0.005901030835625318
ndcg_cut_2,0.0030642572721937857
ndcg_cut_5,0.0033847147595091375
ndcg_cut_10,0.0044111178601985756
map_cut_2,0.0013902864426108535
map_cut_5,0.001974650238888515
map_cut_10,0.002327407439351495
aucroc,0.5981885341516999
![f1 train_valid_loss](https://user-images.githubusercontent.com/124281176/222866091-28a2a8f1-cd8a-46c2-8314-c1745229f217.png)


------------------------------------------------------------------------
Model Layers: [128, 512]
epochs: 60
fold: 1
Positive Weight: 700

,mean
P_2,0.002918092171463071
P_5,0.002844301334943317
P_10,0.0025021801838062767
recall_2,0.0016427245749040871
recall_5,0.004172855833149662
recall_10,0.007397930662168923
ndcg_cut_2,0.0030623060751689616
ndcg_cut_5,0.003724937768290848
ndcg_cut_10,0.0052121740627079
map_cut_2,0.0013406932505342648
map_cut_5,0.002079011564823599
map_cut_10,0.0026002920044310034
aucroc,0.6229254703652766
![f1 train_valid_loss](https://user-images.githubusercontent.com/124281176/222866172-3deffba7-4b0b-4bbe-8768-12cc41bfdc67.png)

------------------------------------------------------------------------
Model Layers: [128, 512]
epochs: 60
fold: 0
Positive Weight: 700

,mean
P_2,0.0033876702220433353
P_5,0.0029248004293285067
P_10,0.0023881398000939248
recall_2,0.0019599293396838162
recall_5,0.0042161400684242335
recall_10,0.006961494599852413
ndcg_cut_2,0.0035318841257492273
ndcg_cut_5,0.00393916644319812
ndcg_cut_10,0.0052423181940984016
map_cut_2,0.0016127769951476928
map_cut_5,0.0023734561689884716
map_cut_10,0.0028378858978577305
aucroc,0.6049406814704197
![f0 train_valid_loss](https://user-images.githubusercontent.com/124281176/222866317-deaa4405-f9e9-4ee5-a306-cc9bdf225e4f.png)

**Comment by hosseinfani** [comment]
@hamedwaezi01-zz 
Thanks. 
- Please report the figure on average of folds. 
- Also, make the loss range fix for all figures. 
- Also, share the codeline that you change the loss. 
- Also, share your findings/thoughts based on your observation of results.

Thank you.

**Comment by hamedwaezi01** [comment]
I will post the average figures of all folds in each session in this comment. Though since the range of losses are different It may not make sense to put them all together. In the next comment I will argue about the results.

positive weight: 50
Layers: [128, 512]
![averaged_train_valid_loss](https://user-images.githubusercontent.com/124281176/223843979-2ac551e4-0de3-40fe-bec9-6c093262f466.png)

#######
positive weight: 50
Layers: [128]
![averaged_train_valid_loss](https://user-images.githubusercontent.com/124281176/223844084-2423b849-c58a-4239-bdc4-f5b5af762178.png)

#######
positive weight: 300
Layers: [128, 512]
![averaged_train_valid_loss](https://user-images.githubusercontent.com/124281176/223844151-e1284ac9-037e-4d00-a3b5-b4b29b1ebed4.png)

#######
positive weight: 300
Layers: [128]
![averaged_train_valid_loss](https://user-images.githubusercontent.com/124281176/223844215-970c7e7e-61fd-407d-93e3-ec24a921c6fc.png)

#######
positive weight: 700
Layers: [128, 512]
![averaged_train_valid_loss](https://user-images.githubusercontent.com/124281176/223844275-48159c9c-3ec6-4eb8-9494-e0e2d8c49887.png)


#######
positive weight: 700
Layers: [128]
![averaged_train_valid_loss](https://user-images.githubusercontent.com/124281176/223844343-3f9e5267-9d3e-4cf5-a968-0ed09b4d977f.png)

**Comment by hamedwaezi01** [comment]
I rerun the models on single layer and noticed that all of them were overfitted before reaching a better loss.
Also I should mention that I added a 0.5 dropout to force the model to generalize.

The bug of zeros was not actually because of the code but because the model hyperparameters were not properly adjusted.
By changing the hyperparameters such as number of layers and increasing the positive weight of the loss, we can see some improvements (an ROC of 0.6 in some cases).

My argument is that we need to fine tune the model itself, loss function, probably the optimizer and also the scheduler. The scheduler tries to adjust (decrease) the LR when needed. Currently we use ReduceLROnPlateau but its inputs should be tuned according to our loss function.
An important point here is that all these hyperparameters affect each other and we might be forced to do some grid search for the best setup.


Some ideas about the positive weight of loss function:
As you can see a large positive weight (700 for DBLP) may improve the metrics but it means that we force the model to give more labels of one. 
And a low positive weight (4 or 3 for DBLP) would lead the model to only output zeros.
This hyperparameter should be set according to the given feature set. The characteristics of DBLP dataset are different than those of IMDB or USPT.

**Comment by hosseinfani** [comment]
#281
